File: src/main/java/linkedlist/AddTwoNumbers.java
package linkedlist;

import structs.ListNode;

public class AddTwoNumbers {
  public static void main(String[] args){
    AddTwoNumbers s = new AddTwoNumbers();
    ListNode num1 = new ListNode(2, new ListNode(4, new ListNode(3)));
    ListNode num2 = new ListNode(5, new ListNode(6, new ListNode(4)));
    ListNode res = s.addTwoNumbers(num1, num2);
    while(res != null){
      System.out.print(res.val + " ");
      res = res.next;
    }
    System.out.print("\n");
  }

  public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    ListNode dummy = new ListNode(-1);
    ListNode copy = dummy;
    int carry = 0;
    while(l1 != null || l2 != null){
      int sum = 0;
      if(l1 != null){
        sum += l1.val;
        l1 = l1.next;
      }

      if(l2 != null){
        sum += l2.val;
        l2 = l2.next;
      }

      sum += carry;
      copy.next = new ListNode(sum%10);
      copy = copy.next;

      carry = sum/10;
    }
    if(carry != 0){
      copy.next = new ListNode(carry);
    }
    return dummy.next;
  }
}

File: src/AddTwoNumbers.kt
import structs.ListNode

fun main(){
    val l1 : ListNode? = ListNode(2, ListNode(4, ListNode(3, null)))
    val l2 : ListNode? = ListNode(5, ListNode(6, ListNode(4, null)))
    var res = addTwoNumbers(l1, l2)
    while(res != null){
        print("${res.`val`} ")
        res = res.next
    }
    print("\n")
}

fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
    var dummy : ListNode? = ListNode(-1, null)
    val copy = dummy
    var carry : Int = 0
    var h1 = l1;
    var h2 = l2
    while(h1 != null || h2 != null){
        var sum = 0
        if(h1 != null){
            sum += h1.`val`
            h1 = h1.next
        }
        if(h2 != null){
            sum += h2.`val`
            h2 = h2.next
        }
        sum += carry
        dummy?.next = ListNode(sum%10)
        dummy = dummy?.next
        carry = sum/10
    }
    if(carry != 0){
        dummy?.next = ListNode(carry)
    }
    return copy?.next
}